### ISSUE TITLE
[MINVOKER-112] duplicate script classpath entries are not filtered out, can cause ClassCastException

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-invoker-plugin#465]
**[luke w patterson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lukewpatterson)** opened **[MINVOKER-112](https://issues.apache.org/jira/browse/MINVOKER-112?redirect=false)** and commented

(it test case attached)

when addTestClassPath http://maven.apache.org/plugins/maven-invoker-plugin/run-mojo.html#addTestClassPath
is set to "true" but the consuming project also depends on groovy or beanshell, scripts fail with ClassCastException,
because the jars end up on the classpath twice

error example:

```
 
groovy.lang.GroovyRuntimeException: Failed to create Script instance for class: class Script1. Reason: java.lang.ClassCastException: Script1 cannot be cast to groovy.lang.GroovyObject
	at org.codehaus.groovy.runtime.InvokerHelper.createScript(InvokerHelper.java:432)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:584)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:564)
	at groovy.lang.GroovyShell.evaluate(GroovyShell.java:542)
	at groovy.lang.GroovyShell.evaluate(GroovyShell.java:518)
	at org.apache.maven.plugin.invoker.GroovyScriptInterpreter.evaluateScript(GroovyScriptInterpreter.java:83)
```

I created a fix locally, but it's rough right now.  Will polish it up and submit later.  Basically, it adds

```
/**
 * @parameter expression="${plugin.artifacts}"
 * @required
 * @readonly
 */
private List pluginArtifacts;
```

to AbstractInvokerMojo, and then spins through the Artifacts, retrieving their file paths, and removes those from the list sent to "scriptRunner.setClassPath"

Granted, the proposed fix could be smarter, cause it only filters out exact matches, but it's better than the current behavior.

---

**Attachments:**
- [script-classpath-duplicates-fix.patch](https://issues.apache.org/jira/secure/attachment/12716682/script-classpath-duplicates-fix.patch) (_1.97 kB_)
- [script-classpath-duplicates-it.patch](https://issues.apache.org/jira/secure/attachment/12716676/script-classpath-duplicates-it.patch) (_11.71 kB_)

1 votes, 0 watchers

**Comment by jira-importer** [comment]
**[luke w patterson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=lukewpatterson)** commented

(attaching fix patch)

Here's a small, clean patch to fix the issue.

It simply removes any class path entries that are duplicated.  Like I mentioned earlier, the logic could probably be smarter, and could try to merge the dependencies using GAV awareness, but this is a good start for now.  This patch fixes situations that would otherwise be clearly wrong (two of same jars on classpath).  I don't think it could cause any passivity issues later if the logic was made more robust.

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

fixed rev 1051652.
Thanks !
